### ISSUE TITLE
Pristine-1.4: Flow accumulates messages that are flushed only at the end.

### ISSUE THREAD

**Issue by kewinrausch** [IRATI/stack#902]
> Whaaat?

If your reaction is like this then you are at the same point as me. This "issue" is visible when you split a flow over multiple Equal cost paths instead of preserving it always on the same one until the end. Actually my policies perform a pick of ECMP ports for a destination in a round-robin style, and so every PDU is sent on a different port (assumed more than one port is available to reach that destination).

I'm using rate based flow with custom policies, which works nicely under normal configuration (all the measurements that has been done until now are correct).

The initial test is done with a classic ECMP policy, which gives me...

**Ping test**

```
krausch@s1:/rping$ ./rping s1 1 dc s4 1 0 10
18384(1461144727)#librina.logs (DBG): New log level: ERR
RTT: 626 us
RTT: 657 us
RTT: 648 us
RTT: 730 us
RTT: 732 us
RTT: 607 us
RTT: 656 us
RTT: 656 us
RTT: 611 us
RTT: 605 us
```

**Bw test**

```
krausch@s4:/rping$ ./rping s4 1 dc --listen
23147(1461145664)#librina.logs (DBG): New log level: ERR
Port 3 allocated for s1-1
Sender:s1, port:3, test:1, recv:6953984, total:6953984, seq:0, ooo: 0, bw:45.876848 Mb/s
Sender:s1, port:3, test:1, recv:12255232, total:19209216, seq:0, ooo: 0, bw:90.109929 Mb/s
Sender:s1, port:3, test:1, recv:12976128, total:32185344, seq:0, ooo: 0, bw:89.177954 Mb/s
Sender:s1, port:3, test:1, recv:13066240, total:45251584, seq:0, ooo: 0, bw:74.633682 Mb/s
Sender:s1, port:3, test:1, recv:16850944, total:62102528, seq:0, ooo: 0, bw:104.722708 Mb/s
Sender:s1, port:3, test:1, recv:15048704, total:77151232, seq:0, ooo: 0, bw:92.888719 Mb/s
Sender:s1, port:3, test:1, recv:11354112, total:88505344, seq:0, ooo: 0, bw:90.473805 Mb/s
Sender:s1, port:3, test:1, recv:13787136, total:102292480, seq:0, ooo: 0, bw:110.407385 Mb/s
Deallocating flow 3...
Sender:s1, port:3, test:1, recv:8740864, total:111033344, seq:0, ooo: 0, bw:69.581440 Mb/s
Flow 3 terminated... out!
Resume for s1[3], recv:112656384, seq:0, out-of-order: 0, bw:86.175437 Mb/s
Servant for port 3 terminated
```

And until here everything goes nicely. When I switch the forwarding policy mechanism to use that kind of distribution (sysfs triggered), then things start to go wrong. I just trigger the policy switch and repeat the experiment...

**Ping test** 

```
krausch@s1:/rping$ ./rping s1 1 dc s4 1 0 10
18699(1461144789)#librina.logs (DBG): New log level: ERR
RTT: 705 us
RTT: 701 us
RTT: 693 us
RTT: 695 us
RTT: 690 us
RTT: 712 us
RTT: 710 us
RTT: 726 us
RTT: 715 us
RTT: 690 us
```

Which is actually correct!

**BW test**

```
Port 3 allocated for s1-1
Sender:s1, port:3, test:1, recv:108544, total:108544, seq:0, ooo: 0, bw:0.714140 Mb/s
Sender:s1, port:3, test:1, recv:4096, total:112640, seq:0, ooo: 0, bw:0.025208 Mb/s
Sender:s1, port:3, test:1, recv:4096, total:116736, seq:0, ooo: 0, bw:0.026425 Mb/s
Sender:s1, port:3, test:1, recv:3072, total:119808, seq:0, ooo: 0, bw:0.024190 Mb/s
Sender:s1, port:3, test:1, recv:3072, total:122880, seq:0, ooo: 0, bw:0.024576 Mb/s
Sender:s1, port:3, test:1, recv:3072, total:125952, seq:0, ooo: 0, bw:0.022504 Mb/s
Sender:s1, port:3, test:1, recv:4096, total:130048, seq:0, ooo: 0, bw:0.025207 Mb/s
Sender:s1, port:3, test:1, recv:3072, total:133120, seq:0, ooo: 0, bw:0.024574 Mb/s
Deallocating flow 3...
Sender:s1, port:3, test:1, recv:3072, total:136192, seq:0, ooo: 0, bw:0.024576 Mb/s
Flow 3 terminated... out!
Resume for s1[3], recv:112566272, seq:0, out-of-order: 0, bw:85.219415 Mb/s
Servant for port 3 terminated
```

Veeery bad!
But let me make you notice of the resume for the experiment:
`Resume for s1[3], recv:112566272, seq:0, out-of-order: 0, bw:85.219415 Mb/s`

My first idea was that the application rping was bugged, and I checked that in order to find the issue. But no issue has been found and all the other tests were good (before the policy switch).

Besides, when looking at the `ifconfig` statistics I can read...

**Sender**

```
eth44.103 Link encap:Ethernet  HWaddr 00:25:90:30:0a:f8
          inet6 addr: fe80::225:90ff:fe30:af8/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:605898 errors:0 dropped:0 overruns:0 frame:0
          TX packets:715474 errors:0 dropped:20 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:25564292 (24.3 MiB)  TX bytes:755722379 (720.7 MiB)

eth47.102 Link encap:Ethernet  HWaddr 00:25:90:30:0a:f9
          inet6 addr: fe80::225:90ff:fe30:af9/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:715322 errors:0 dropped:0 overruns:0 frame:0
          TX packets:605468 errors:0 dropped:10 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:30329979 (28.9 MiB)  TX bytes:639435572 (609.8 MiB)
```

**Receiver**

```
eth44.204 Link encap:Ethernet  HWaddr 00:25:90:32:8d:51
          inet6 addr: fe80::225:90ff:fe32:8d51/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:646521 errors:0 dropped:0 overruns:0 frame:0
          TX packets:756414 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:673695576 (642.4 MiB)  TX bytes:40148092 (38.2 MiB)

eth46.304 Link encap:Ethernet  HWaddr 00:25:90:32:8d:50
          inet6 addr: fe80::225:90ff:fe32:8d50/64 Scope:Link
          UP BROADCAST RUNNING MULTICAST  MTU:1500  Metric:1
          RX packets:756674 errors:0 dropped:0 overruns:0 frame:0
          TX packets:647155 errors:0 dropped:0 overruns:0 carrier:0
          collisions:0 txqueuelen:0
          RX bytes:788591327 (752.0 MiB)  TX bytes:34355010 (32.7 MiB)
```

This tells me that packets ARE actually received by the node, in the same, less or more, amount of the sent one. My next step then has been to measure the server/destination node memory and see what happens during a 20 second test, and the result was:

```
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       703M        10G         0B       107M       410M
-/+ buffers/cache:       185M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -hm
             total       used       free     shared    buffers     cached
Mem:           11G       703M        10G         0B       107M       410M
-/+ buffers/cache:       185M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       702M        10G         0B       107M       410M
-/+ buffers/cache:       184M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       773M        10G         0B       107M       421M
-/+ buffers/cache:       243M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       787M        10G         0B       107M       423M
-/+ buffers/cache:       255M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       799M        10G         0B       107M       425M
-/+ buffers/cache:       266M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       812M        10G         0B       107M       427M
-/+ buffers/cache:       277M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       824M        10G         0B       107M       429M
-/+ buffers/cache:       286M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       836M        10G         0B       107M       431M
-/+ buffers/cache:       296M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       848M        10G         0B       107M       433M
-/+ buffers/cache:       307M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       862M        10G         0B       107M       435M
-/+ buffers/cache:       319M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       874M        10G         0B       107M       437M
-/+ buffers/cache:       328M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       887M        10G         0B       107M       439M
-/+ buffers/cache:       340M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       901M        10G         0B       107M       441M
-/+ buffers/cache:       351M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       915M        10G         0B       107M       443M
-/+ buffers/cache:       363M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       929M        10G         0B       107M       445M
-/+ buffers/cache:       375M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       943M        10G         0B       107M       448M
-/+ buffers/cache:       387M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       957M        10G         0B       107M       450M
-/+ buffers/cache:       399M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       971M        10G         0B       107M       452M
-/+ buffers/cache:       411M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       985M        10G         0B       107M       455M
-/+ buffers/cache:       422M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       999M        10G         0B       107M       457M
-/+ buffers/cache:       434M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       1.0G        10G         0B       107M       459M
-/+ buffers/cache:       444M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       1.0G        10G         0B       107M       460M
-/+ buffers/cache:       449M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       758M        10G         0B       107M       460M
-/+ buffers/cache:       190M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       758M        10G         0B       107M       460M
-/+ buffers/cache:       190M        11G
Swap:         3.0G         0B       3.0G
krausch@s4:~$ free -h
             total       used       free     shared    buffers     cached
Mem:           11G       758M        10G         0B       107M       460M
-/+ buffers/cache:       190M        11G
Swap:         3.0G         0B       3.0G
```

Memory grows a lot during this small experiment (~300 Mb at 90M b/s rate), and this is telling me that somewhere the PDU are accumulated. I don't know why, but it seems that when the flow is de-allocated all the traffic is flushed to the listening application (which is still in listening mode).

You can reach my policies under the pristine-1.4 branch of my fork, https://github.com/kewinrausch/stack/tree/pristine-1.4/plugins/rdsr-ps

**Comment by edugrasa** [comment]
Hi Kewin,

Few questions to make sure I understand the problem well :) This only happens with the combination of rate-based flow control and your ECMP policies? So rate-based flow control + hash-based ECMP works well, right? (it is the first test) 
Have you tried window-based flow control with your ECMP policy? and window-based flow control with hash-based ECMP? I can try to find time to help you with this tests tomorrow.

**Comment by kewinrausch** [comment]
The 1st, successful test, was rate-based flow + a forwarding strategy (the default one) which maintains all the pdus of one flow on the same path.

The 2nd test was rate-based flow + forwarding strategy which uses round-robin style to distribute pdus on all the ECMP available ports.

I can give a try with window-based and the incriminated forwarding strategy. :)
I'll let you know about this.

**Comment by kewinrausch** [comment]
Quick report with window-based flow + forwarding strategy set on Water distribution (option 2 in /sys/pffb/fwd_type).

```
krausch@s1:/rping$ ./rping s1 1 dc s4 1 1 20 12500
5196(1461156681)#librina.logs (DBG): New log level: ERR
Starting experiment 1 to s4, PDU at 1024 bytes...
Port:3, test:1, sent:9569280, seq:9345, bw:76.614229 Mb/s
Port:3, test:1, sent:21095424, seq:29946, bw:141.299325 Mb/s
Port:3, test:1, sent:9534464, seq:39257, bw:76.344804 Mb/s
Port:3, test:1, sent:11237376, seq:50231, bw:83.484401 Mb/s
Port:3, test:1, sent:3686400, seq:53831, bw:28.029970 Mb/s
Port:3, test:1, sent:2439168, seq:56213, bw:19.532056 Mb/s
Port:3, test:1, sent:4331520, seq:60443, bw:32.180235 Mb/s
Port:3, test:1, sent:9473024, seq:69694, bw:75.853067 Mb/s
Port:3, test:1, sent:8509440, seq:78004, bw:54.156705 Mb/s
Port:3, test:1, sent:1711104, seq:79675, bw:10.795439 Mb/s
Port:3, test:1, sent:1912832, seq:81543, bw:12.031657 Mb/s
Port:3, test:1, sent:3645440, seq:85103, bw:27.836366 Mb/s
Port:3, test:1, sent:1843200, seq:86903, bw:11.588645 Mb/s
Port:3, test:1, sent:8532992, seq:95236, bw:54.355547 Mb/s
Port:3, test:1, sent:5704704, seq:100807, bw:40.173229 Mb/s
Port:3, test:1, sent:2497536, seq:103246, bw:19.995565 Mb/s
Port:3, test:1, sent:11374592, seq:114354, bw:84.210467 Mb/s
Port:3, test:1, sent:1879040, seq:116189, bw:11.780963 Mb/s
Port:3, test:1, sent:635904, seq:116810, bw:4.184116 Mb/s
Port:3, test:1, sent:530432, seq:117328, bw:3.477510 Mb/s
Releasing the flow on 3...
Resume for 3: sent:120143872, seq:117328, bw:42.346236 Mb/s
```

The link has been limited to 100 Mb/s with ethtool, so these performances are a little strange... in addition of this, looking at the memory, I got:

**First experiment**
```
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        409      11471          0         18        231
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        412      11468          0         18        235
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        413      11467          0         18        236
-/+ buffers/cache:        158      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        415      11465          0         18        239
-/+ buffers/cache:        157      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        416      11465          0         18        240
-/+ buffers/cache:        157      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        421      11459          0         18        240
-/+ buffers/cache:        162      11718
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        421      11459          0         18        243
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        427      11453          0         18        249
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        429      11451          0         18        252
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        430      11450          0         18        253
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        431      11449          0         18        254
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        432      11448          0         18        255
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        439      11441          0         18        263
-/+ buffers/cache:        157      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        441      11439          0         18        264
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        442      11438          0         18        265
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        444      11436          0         18        267
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        447      11434          0         18        269
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        447      11433          0         18        270
-/+ buffers/cache:        158      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        447      11433          0         18        271
-/+ buffers/cache:        157      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        448      11432          0         18        271
-/+ buffers/cache:        158      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        448      11432          0         18        271
-/+ buffers/cache:        158      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        447      11433          0         18        271
-/+ buffers/cache:        157      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        448      11432          0         18        271
-/+ buffers/cache:        158      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        448      11433          0         18        271
-/+ buffers/cache:        157      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        448      11433          0         18        271
-/+ buffers/cache:        157      11723
Swap:         3071          0       3071
```

**Second experiment**
```
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        460      11420          0         18        283
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        461      11419          0         18        284
-/+ buffers/cache:        159      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        462      11418          0         18        285
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        465      11416          0         18        287
-/+ buffers/cache:        158      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        466      11415          0         18        289
-/+ buffers/cache:        158      11723
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        467      11413          0         18        289
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        468      11412          0         18        290
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        469      11411          0         18        291
-/+ buffers/cache:        159      11722
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        472      11408          0         18        294
-/+ buffers/cache:        160      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        472      11408          0         18        294
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        476      11405          0         18        297
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        476      11405          0         18        297
-/+ buffers/cache:        160      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        476      11404          0         18        298
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        476      11405          0         18        298
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        476      11404          0         18        298
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
krausch@s4:~$ free -m
             total       used       free     shared    buffers     cached
Mem:         11881        476      11404          0         18        298
-/+ buffers/cache:        159      11721
Swap:         3071          0       3071
```

The memory is always loaded and does not seems to be released.... Still this happens at a "lower" rate than with rate based. 

**An additional note:** take in account that I also have qos with **ordered** option set at false.

**Comment by edugrasa** [comment]
Fixed by #920
